proposals/0025-shader-execution-reordering.md

```diff
@@ -0,0 +1,1519 @@
+execution of traversal, intersection testing, anyhit shading and closesthit or
+miss shading imposes various restrictions on the programming model that, again,
+can increase the amount of developer effort and decrease performance. One
+aspect is that common code, e.g., vertex fetch and interpolation, must be
+duplicated in all closesthit shaders. This can cause more code to be generated,
+which is particularly problematic in a divergent execution environment.
+Furthermore, `TraceRay`'s nature requires that simple visibility rays
+unnecessarily execute hit shaders in order to access basic information about
+the hit which must be transferred back to the caller through the payload.
+
+## Proposed Solution
+
+Shader Execution Reordering (SER) introduces a new HLSL built-in intrinsic,
+`ReorderThread`,
+that enables application-controlled reordering of work across the GPU for
+improved execution and data coherence.
+Additionally, the introduction of `HitObject` allows separation of traversal,
+anyhit shading and intersection testing from closesthit and miss shading.
+
+`HitObject` and `ReorderThread` can be combined to improve coherence for
+closesthit and miss shader execution in a controlled manner.
+Applications can control coherence based on hit properties,
+ray generation state, ray payload, or any combination thereof. Applications can
+maximize performance by considering coherence for both hit and miss shading as
+well as subsequent control flow and data access patterns inside the
+raygeneration shader.
+
+`HitObject` improves the flexibility of the ray tracing pipeline in general.
+First, common code, such as vertex fetch and interpolation, must no longer be
+duplicated in all closesthit shaders. Common code can simply be part of the
+raygeneration shader and execute before closesthit shading. Second, simple
+visibility rays no longer have to invoke hit shaders in order to access basic
+information about the hit, such as the distance to the closest hit. Finally,
+`HitObject` can be constructed from a `RayQuery`, which enables
+`ReorderThread` and shader table-based closesthit and miss shading to be combined with
+`RayQuery`.
+
+The proposed extension to HLSL should be relatively straightforward to adopt by
+current DXR implementations: `HitObject` merely decouples existing `TraceRay`
+functionality into two distinct stages: the traversal stage and the shading
+stage.
+For SER's `ReorderThread`, the minimal allowed implementation is simply a
+no-op, while implementations that already employ more sophisticated scheduling
+strategies are likely able to reuse existing mechanisms to implement support
+for SER. No DXR runtime changes are necessary, since the proposed extension to
+the programming model is limited to HLSL and DXIL.
+
+## Detailed Design
+
+This section describes the HLSL additions for `HitObject` and `ReorderThread`
+in detail.
+The canonical use of these features involve changing a `TraceRay` call to the
+following sequence:
```

Comment:
we've spoken last time about this being equivalent functionally.
The current sentence "almost" says that. Could we have this more explicitly?
```suggestion
The canonical use of these features involve changing a `TraceRay` call to the
following sequence that is functionally equivalent:
```